docs/topic-guides/elastic-infra/aws-spot.txt

```diff
@@ -0,0 +1,80 @@
+.. _aws-spot:
+
+####################
+ AWS Spot Instances
+####################
+
+This document describes how AWS spot instances work. Determined handles
+most details of working with spot instances, but knowing how spot
+instances work at a high level is helpful for understanding the behavior
+of Determined when provisioning spot instances.
+
+*******************
+ On-Demand vs Spot
+*******************
+
+The standard way to create an EC2 instance is to create an on-demand
+instance. On-demand instances always have the same price and once you
+have created one, you are guaranteed to have that instance forever
+(barring hardware failure).
+
+The number of EC2 instances desired by customers varies over time. AWS
+wants to have enough hardware so that during peak usage periods, all
+customers are able to get as many on-demand instances as they need. This
+means that during non-peak periods (which is most of the time), AWS has
+extra hardware capacity sitting around unused. AWS makes these extra
+instances available to customers via the spot market as spot instances.
+They can be rented at a much reduced cost, but the trade-off is that, if
+AWS needs instances to satisfy on-demand users, they can reclaim your
+spot instance and give it to the on-demand user. If all available
+instances are in use by on-demand customers, you will not be able to
+launch any spot instances.
+
+********************
+ Spot On Determined
+********************
+
+Because they can be reclaimed, using spot instances requires you to have
+good fault tolerance built into your software. Determined was built with
+fault-tolerance as a core feature, so using spot instances is usually as
+easy as setting ``spot: true`` in the master configuration.
+
+However, Determined cannot solve the problem of AWS not having enough
+capacity. If you are using Determined with spot instances and are
+expecting instances to be created, but you aren't seeing them get
+created, it is possible that Determined is sending requests to AWS to
+create spot instances, but AWS is out of capacity so the requests are
+stuck until capacity becomes available. Unfortunately, AWS's available
+capacity at any time is a tightly-guarded secret, so the best Determined
+can do is request instances and see how AWS responds. If Determined is
+unable to create spot instances due to lack of AWS capacity, this will
+be visible in the master logs.
```

Comment:
```suggestion
However, AWS might not always have the capacity for spot requests.
When AWS is out of capacity, the Determined cluster would wait until
the spot requests are fulfilled. This will be visible in the master logs.
However, you can increase the max spot price to let the spot requests 
more likely to be fulfilled.
```

This sounds more clear what users should care about and do to react. What do you think?

Reply:
I'm not against improving the text, but I don't think this is very clear. What about the original text do you think we should improve?

If there are capacity issues, change the max spot price will have no impact.

Reply:
> If there are capacity issues, changing the max spot price will have no impact.

I mean capacity by capacity with a specific max spot price but the last sentence I propose is not precise.

I think the original text is more about what Determined solve and not solve so it is not very friendly to users I think. But what the users care about is how they identify the issues and how they react to them to resolve the issues.

> However, Determined cannot solve the problem of AWS not having enough capacity. If you are using Determined with spot instances and are expecting instances to be created, but you aren't seeing them get created, it is possible that Determined is sending requests to AWS to create spot instances, but AWS is out of capacity so the requests are stuck until capacity becomes available. Unfortunately, AWS's available capacity at any time is a tightly-guarded secret, so the best Determined can do is request instances and see how AWS responds. If Determined is unable to create spot instances due to lack of AWS capacity, this will be visible in the master logs.

So basically this paragraph is saying: when AWS doesn't have the capacity, users are able to observe it in the logs and should just wait. 

> it is possible that Determined is sending requests to AWS to create spot instances, but AWS is out of capacity so the requests are stuck until capacity becomes available

This might not be very precise. It could be the AWS is starting up the instances or the provisioner decides not to start up due to maximum instances. For those two cases, users should do differently from AWS not having the capacity. So I think we should avoid using such term "possible" because it is not explicit in what users should do.

> Unfortunately, AWS's available capacity at any time is a tightly-guarded secret, so the best Determined can do is request instances and see how AWS responds.

Even if the available capacity is not a secret, the master would still just wait for the requests to be fulfilled. So I think the original text to explain this should be just removed.

Reply:
Updated